File: inheritance.cpp
#include <iostream>
using namespace std;

// Base class
class Square {
protected:
    double side;

public:
    // Parameterized constructor
    Square(double s) {
        side = s;
        cout << "Square constructor called" << endl;
    }

    // Member function to get perimeter
    double getPeri() {
        return 4 * side;
    }

    // Member function to get area
    double getArea() {
        return side * side;
    }

    // Destructor
    ~Square() {
        cout << "Square destructor called" << endl;
    }
};

// Derived class
class Cube : public Square {
public:
    // Constructor
    Cube(double s) : Square(s) {
        cout << "Cube constructor called" << endl;
    }

    // Member function to get surface area of cube
    double getArea() {
        return 6 * side * side;
    }

    // Member function to get volume of cube
    double getVolume() {
        return side * side * side;
    }

    // Destructor
    ~Cube() {
        cout << "Cube destructor called" << endl;
    }
};

// Main function
int main() {
    Square sq(5);
    cout << "Square Perimeter: " << sq.getPeri() << endl;
    cout << "Square Area: " << sq.getArea() << endl;

    cout << endl;

    Cube cb(5);
    cout << "Cube Surface Area: " << cb.getArea() << endl;
    cout << "Cube Volume: " << cb.getVolume() << endl;

    return 0;
}

File: inheritance.py
# Base class
class Square:
    def __init__(self, side):
        self.side = side
        print("Square constructor called")

    def get_peri(self):
        return 4 * self.side

    def get_area(self):
        return self.side * self.side

    def __del__(self):
        print("Square destructor called")


# Derived class
class Cube(Square):
    def __init__(self, side):
        super().__init__(side)
        print("Cube constructor called")

    # Override get_area (surface area of cube)
    def get_area(self):
        return 6 * self.side * self.side

    def get_volume(self):
        return self.side * self.side * self.side

    def __del__(self):
        print("Cube destructor called")


# Testing the classes
if __name__ == "__main__":
    sq = Square(4)
    print("Square Perimeter:", sq.get_peri())
    print("Square Area:", sq.get_area())

    print()

    cb = Cube(4)
    print("Cube Surface Area:", cb.get_area())
    print("Cube Volume:", cb.get_volume())
